spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java

```diff
@@ -407,15 +409,37 @@ private Builder doExecuteWithPartialProgress(
 
   Stream<RewriteFileGroup> toGroupStream(
       RewriteExecutionContext ctx, Map<StructLike, List<List<FileScanTask>>> groupsByPartition) {
-    return groupsByPartition.entrySet().stream()
+
+    List<RewriteFileGroup> selectedFileGroups = Lists.newArrayList();
+    AtomicInteger fileCountRunner = new AtomicInteger();
+
+    groupsByPartition.entrySet().stream()
         .filter(e -> !e.getValue().isEmpty())
-        .flatMap(
-            e -> {
-              StructLike partition = e.getKey();
-              List<List<FileScanTask>> scanGroups = e.getValue();
-              return scanGroups.stream().map(tasks -> newRewriteGroup(ctx, partition, tasks));
-            })
-        .sorted(RewriteFileGroup.comparator(rewriteJobOrder));
+        .forEach(
+            entry -> {
+              StructLike partition = entry.getKey();
+              entry
+                  .getValue()
+                  .forEach(
+                      fileScanTasks -> {
+                        if (maxFilesToRewrite == null) {
+                          selectedFileGroups.add(newRewriteGroup(ctx, partition, fileScanTasks));
+                        } else if (fileCountRunner.get() < maxFilesToRewrite) {
+                          int remainingSize = maxFilesToRewrite - fileCountRunner.get();
+                          int fileScanTaskListSize = fileScanTasks.size();
+                          selectedFileGroups.add(
+                              newRewriteGroup(
+                                  ctx,
+                                  partition,
+                                  remainingSize >= fileScanTaskListSize
+                                      ? fileScanTasks
+                                      : fileScanTasks.subList(0, remainingSize)));
+                          fileCountRunner.getAndAdd(Math.min(remainingSize, fileScanTaskListSize));
```

Comment:
I think we can avoid the terinary operator above and some of the complexity here by just calculating this Min above, and always calling subList(0, minValue)

Reply:
Sure ! It was a miss from my end to assume that sublist is not  O(1) complexity . I did remove the terinary operator , pre computed the variable to update the runner and also extract sublist accordingly.